File: r-package/R/pipeline_nodes.R
#' Validate a single DAG entry from the JSON structure
#'
#' Validates that the entry is a list/object, extracts the node name and
#' its list of dependencies, checking that they are non-empty strings.
#'
#' @param entry A list representing a single node's configuration from the DAG JSON.
#' @param index Integer. The 1-based index of this entry in the DAG file (used in error messages).
#' @param dag_file Character. The path to the DAG file being parsed (used in error messages).
#'
#' @return A list containing:
#'   \item{node_name}{The validated, non-empty character string node name.}
#'   \item{depends}{A sorted, unique character vector of dependency node names.}
#'
#' @keywords internal
validate_node_entry <- function(entry, index, dag_file) {
  if (!is.list(entry)) {
    stop(
      sprintf("Entry %d in `%s` must be an object.", index, dag_file),
      call. = FALSE
    )
  }

  node_name <- entry$node_name
  depends <- entry$depends

  if (is.list(depends)) {
    depends <- unlist(depends)
  }

  if (is.null(depends)) {
    depends <- character(0)
  }

  if (!is.character(node_name) || length(node_name) != 1L || is.na(node_name) || !nzchar(node_name)) {
    stop(
      sprintf("Entry %d in `%s` has an invalid `node_name`.", index, dag_file),
      call. = FALSE
    )
  }

  if (!is.character(depends) || any(is.na(depends)) || any(!nzchar(depends))) {
    stop(
      sprintf("Node `%s` in `%s` has an invalid `depends` list.", node_name, dag_file),
      call. = FALSE
    )
  }

  list(node_name = node_name, depends = unique(sort(depends)))
}

#' Get pipeline nodes and their dependencies from `_pipeline/dag.json`
#'
#' Reads a pipeline DAG definition from `_pipeline/dag.json` and returns a
#' data frame with node names and their dependencies.
#'
#' @param pipeline_dir Path to the pipeline build directory. Defaults to
#'   `"_pipeline"`.
#' @param dag_file Name of the DAG JSON file inside `pipeline_dir`.
#'   Defaults to `"dag.json"`.
#'
#' @return A data frame with two columns: `node` (character) and `depends`
#'   (list of character vectors).
#'
#' @details
#' The function reads the JSON file, parses its array of node definitions,
#' validates that all dependencies refer to existing nodes defined within the same
#' DAG file, and that no duplicate node names exist.
#'
#' @examples
#' \dontrun{
#'   # Assuming a pipeline has been constructed in "_pipeline"
#'   nodes <- pipeline_nodes()
#' }
#'
#' @export
pipeline_nodes <- function(pipeline_dir = "_pipeline", dag_file = "dag.json") {
  validate_scalar_string(pipeline_dir, "pipeline_dir")
  validate_scalar_string(dag_file, "dag_file")

  if (!dir.exists(pipeline_dir)) {
    stop(sprintf("Pipeline directory `%s` does not exist.", pipeline_dir), call. = FALSE)
  }

  dag_path <- file.path(pipeline_dir, dag_file)
  if (!file.exists(dag_path)) {
    stop(sprintf("DAG file `%s` does not exist.", dag_path), call. = FALSE)
  }

  dag <- tryCatch(
    jsonlite::fromJSON(dag_path, simplifyVector = FALSE),
    error = function(err) {
      stop(sprintf("Failed to read DAG file `%s`: %s", dag_path, conditionMessage(err)), call. = FALSE)
    }
  )

  if (!is.list(dag)) {
    stop(sprintf("DAG file `%s` must decode to an array.", dag_path), call. = FALSE)
  }

  normalized <- lapply(seq_along(dag), function(i) validate_node_entry(dag[[i]], i, dag_path))
  node_names <- vapply(normalized, `[[`, character(1), "node_name")

  if (any(duplicated(node_names))) {
    duplicated_names <- unique(node_names[duplicated(node_names)])
    stop(sprintf("DAG file `%s` has duplicate node_name values: %s", dag_path, paste(duplicated_names, collapse = ", ")), call. = FALSE)
  }

  unknown_dependencies <- unique(unlist(lapply(normalized, function(entry) setdiff(entry$depends, node_names)), use.names = FALSE))
  if (length(unknown_dependencies) > 0L) {
    stop(sprintf("DAG file `%s` references unknown dependencies: %s", dag_path, paste(unknown_dependencies, collapse = ", ")), call. = FALSE)
  }

  data.frame(
    node = node_names,
    depends = I(lapply(normalized, `[[`, "depends")),
    stringsAsFactors = FALSE
  )
}

File: py-package/src/tlang/pipeline_nodes.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any


def _validate_non_empty_string(value: str, arg: str) -> None:
    """Validate that a value is a non-empty, stripped string.

    Parameters
    ----------
    value : str
        The value to check.
    arg : str
        The name of the argument being checked (used in error messages).

    Raises
    ------
    ValueError
        If `value` is not a string, or if it consists solely of whitespace.
    """
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"`{arg}` must be a non-empty string.")


def _validate_entry(entry: Any, index: int, dag_path: Path) -> tuple[str, list[str]]:
    """Validate a single DAG entry dictionary from the JSON structure.

    Extracts the node name and its dependency list, ensuring all entries are valid.

    Parameters
    ----------
    entry : Any
        The individual entry object from the DAG file, expected to be a dictionary.
    index : int
        The 1-based index of this entry in the DAG file (used in error messages).
    dag_path : Path
        The Path to the DAG file being parsed (used in error messages).

    Returns
    -------
    tuple[str, list[str]]
        A tuple containing:
        - The validated node name (as a non-empty string).
        - A sorted, duplicate-free list of dependency node names (as strings).

    Raises
    ------
    ValueError
        If the entry is not a dictionary, if `node_name` is missing or invalid,
        or if `depends` is not a list of non-empty strings.
    """
    if not isinstance(entry, dict):
        raise ValueError(f"Entry {index} in `{dag_path}` must be an object.")

    node_name = entry.get("node_name")
    depends = entry.get("depends")
    if depends is None:
        depends = []

    if not isinstance(node_name, str) or not node_name.strip():
        raise ValueError(f"Entry {index} in `{dag_path}` has an invalid `node_name`.")

    if not isinstance(depends, list) or any(not isinstance(dep, str) or not dep.strip() for dep in depends):
        raise ValueError(f"Node `{node_name}` in `{dag_path}` has an invalid `depends` list.")

    return node_name, sorted(set(depends))


def pipeline_nodes(pipeline_dir: str | Path = "_pipeline", dag_file: str = "dag.json") -> dict[str, list[str]]:
    """Get pipeline nodes and their dependencies from the DAG configuration.

    Reads and validates the DAG definition from a JSON file (typically `_pipeline/dag.json`)
    and returns a mapping of node names to their lists of dependencies.

    Parameters
    ----------
    pipeline_dir : str | Path, optional
        The path to the pipeline directory where the DAG file is located.
        Defaults to "_pipeline".
    dag_file : str, optional
        The filename of the DAG configuration. Defaults to "dag.json".

    Returns
    -------
    dict[str, list[str]]
        A dictionary mapping each node name (str) to the list of node names
        it depends on (list of str), sorted alphabetically.

    Raises
    ------
    ValueError
        If `pipeline_dir` or `dag_file` is invalid/empty, if the JSON structure is
        malformed, contains duplicate node names, or references unknown dependencies.
    FileNotFoundError
        If the pipeline directory or the DAG file does not exist.
    OSError
        If the DAG file exists but cannot be read due to an I/O error.
    """
    pipeline_path = Path(pipeline_dir)
    _validate_non_empty_string(str(pipeline_path), "pipeline_dir")
    _validate_non_empty_string(dag_file, "dag_file")

    if not pipeline_path.is_dir():
        raise FileNotFoundError(f"Pipeline directory `{pipeline_path}` does not exist.")

    dag_path = pipeline_path / dag_file
    if not dag_path.is_file():
        raise FileNotFoundError(f"DAG file `{dag_path}` does not exist.")

    try:
        with dag_path.open("r", encoding="utf-8") as handle:
            data = json.load(handle)
    except OSError as err:
        raise OSError(f"Failed to read DAG file `{dag_path}`: {err}") from err
    except json.JSONDecodeError as err:
        raise ValueError(f"Failed to read DAG file `{dag_path}`: {err}") from err

    if not isinstance(data, list):
        raise ValueError(f"DAG file `{dag_path}` must decode to an array.")

    normalized = [_validate_entry(entry, idx + 1, dag_path) for idx, entry in enumerate(data)]
    node_names = [name for name, _ in normalized]

    duplicates = sorted({name for name in node_names if node_names.count(name) > 1})
    if duplicates:
        raise ValueError(
            f"DAG file `{dag_path}` has duplicate node_name values: {', '.join(duplicates)}"
        )

    node_set = set(node_names)
    unknown_deps = sorted({dep for _, deps in normalized for dep in deps if dep not in node_set})
    if unknown_deps:
        raise ValueError(
            f"DAG file `{dag_path}` references unknown dependencies: {', '.join(unknown_deps)}"
        )

    return dict(normalized)
